mozilla/fx-private-relay-add-on
# Change icon: overlaid button to background image

This PR fixes #22, #80, #227, #253, #360, #358, #394, #397, #400, #429, #447, #454, #497, #509.

(For anyone coming across this PR: we'll still need to do some extensive testing to make sure this doesn't break commonly-used websites, so please be patient — it may still take a while to land.)

# New feature description

This applies a new method of injecting the Relay icon that I prototyped a while ago; a helpful description of this method can be found [here](https://github.com/mozilla/fx-private-relay-add-on/issues/22#issuecomment-1372445245), by @czlucius who independently landed on the same approach.

With this method, we don't have to change the stacking order of the website, and thus should avoid a lot of issues where we break the website we're injecting in. It did require some extra work to stay accessible; we still inject a button that is visually hidden, but accessible to screen reader users and people who navigate using the keyboard. I've also modified the hover/focus state styling to make it easier. Scrolling with the menu open is no longer possible; I can probably re-enable that with a re-positioning scroll listener, but I feel like the user experience is probably fine if you dismiss the menu first, and should be better for performance.

Note that I carried over the code in src/js/other-websites/icon.js from the experimental extension I prototyped this in, so the code style might be a little different from the rest of the codebase - the experimental extension had TypeScript and helpers I've to rip out.

# Screenshot (if applicable)

![image](https://user-images.githubusercontent.com/4251/215064034-87013994-837e-4c32-8dc8-dddc38ca1e35.png)

[Screencast from 2023-01-27 11-26-42.webm](https://user-images.githubusercontent.com/4251/215064658-b9d7c905-7bca-4d28-af69-275e0a45120e.webm)

![image](https://user-images.githubusercontent.com/4251/215064576-63e90c84-7536-4a7f-9228-0ff2692b9500.png)

![image](https://user-images.githubusercontent.com/4251/215064627-e0b10638-2d66-4276-99df-49af4570fc62.png)

![image](https://user-images.githubusercontent.com/4251/215065032-7a9d8efb-bdbb-4e30-bc69-6354b3a53375.png)

![image](https://user-images.githubusercontent.com/4251/215065333-531a1177-e804-4b12-898e-2406ac28732e.png)

![image](https://user-images.githubusercontent.com/4251/215065428-364b07e7-2a1c-48da-9693-cc1917191cf5.png)

![image](https://user-images.githubusercontent.com/4251/215065521-4fe0bbb1-c026-467f-b8a6-74946567e243.png)

![image](https://user-images.githubusercontent.com/4251/215065696-3281cdad-3e76-4d8c-be83-417a9e91de8c.png)

![image](https://user-images.githubusercontent.com/4251/215065767-56f2a538-e25c-4889-812b-bc41d19c5412.png)

# How to test

Visit websites with the Relay icon, and make sure that everything looks and works OK.

# Checklist
